src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocInitializerTests.cs

```diff
@@ -5,13 +5,101 @@
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
 using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
 using Xunit;
 
 namespace Microsoft.CodeAnalysis.CSharp.UnitTests
 {
     [CompilerTrait(CompilerFeature.StackAllocInitializer)]
     public class StackAllocInitializerTests : CompilingTestBase
     {
+        [Fact, WorkItem(33945, "https://github.com/dotnet/roslyn/issues/33945")]
+        public void RestrictedTypesAllowedInStackalloc()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+public ref struct RefS { }
+public ref struct RefG<T> { public T field; }
+
+class C
+{
+    unsafe void M()
+    {
+        var x1 = stackalloc RefS[10];
+        var x2 = stackalloc RefG<string>[10];
+        var x3 = stackalloc RefG<int>[10];
+        var x4 = stackalloc System.TypedReference[10];
```

Comment:
TypedReference has a byref field in it (logically, similar to `Span<T>`). I do not think stackallocs are reported to GC. 
If this is the only ref, it could result in a GC hole.

Reply:
ArgIterator is probably ok. Also has a ref, but I think it always points to unmovable locations.

Reply:
I wonder what happens with `Span<int>`  - do we get an error because it is not unmanaged?

Reply:
Correct. This will change once this gets merged to master (which has the "unmanaged generic struct" feature). Jared pointed out that the test expectations will need updated during the merge.

---
In reply to: [263615857](https://github.com/dotnet/roslyn/pull/33951#discussion_r263615857) [](ancestors = 263615857)

Reply:
Just being ref-like should not prevent stackalloc, but some of them are not technically "unmanaged" since they may contain embedded references to movable heap. Are we able to sort out such distinction by looking at metadata?

Reply:
Interestingly we currently explicitly assume that TypedReference is unmanaged. For the purpose of taking a pointer to a variable of such type it seems ok. - They cannot be on the heap, so cannot move...

We may need a slightly different criteria for stackalloc, since now the concern is what variables themselves may refer to. 
It might be a problem only for a few "magic" types.

Reply:
From discussion with Jared, Andy and David: you should be able to stackalloc any unmanaged type. 
`Span` is considered managed, and has a dummy `object` field in the various ref assemblies.
`TypedReference` currently does not have such a dummy `object` field. Filed corefx issue [https://github.com/dotnet/corefx/issues/35873](https://github.com/dotnet/corefx/issues/35873)

---
In reply to: [263623732](https://github.com/dotnet/roslyn/pull/33951#discussion_r263623732) [](ancestors = 263623732)

Reply:
I think the unmanagedness of TypedReference is hardcoded in the compiler. 

` TypedReference* ` has always been legal.  If corefx adds a dummy object and if compiler starts looking at that, you will break that.